core/capabilities/webapi/target/capability.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/smartcontractkit/chainlink-common/pkg/types/core"
+	"github.com/smartcontractkit/chainlink-common/pkg/values"
+	"github.com/smartcontractkit/chainlink/v2/core/capabilities/validation"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/handlers/webcapabilities"
+)
+
+const ID = "web-api-target@1.0.0"
+
+var _ capabilities.TargetCapability = &Capability{}
+
+var capabilityInfo = capabilities.MustNewCapabilityInfo(
+	ID,
+	capabilities.CapabilityTypeTarget,
+	"A target that sends HTTP requests to external clients via the Chainlink Gateway.",
+)
+
+// Capability is a target capability that sends HTTP requests to external clients via the Chainlink Gateway.
+type Capability struct {
+	capabilityInfo        capabilities.CapabilityInfo
+	connectorHandler      *ConnectorHandler
+	lggr                  logger.Logger
+	registry              core.CapabilitiesRegistry
+	config                Config
+	registeredWorkflows   map[string]struct{}
+	registeredWorkflowsMu sync.Mutex
+}
+
+func NewCapability(config Config, registry core.CapabilitiesRegistry, connectorHandler *ConnectorHandler, lggr logger.Logger) (*Capability, error) {
+	return &Capability{
+		capabilityInfo:        capabilityInfo,
+		config:                config,
+		registry:              registry,
+		connectorHandler:      connectorHandler,
+		registeredWorkflows:   make(map[string]struct{}),
+		registeredWorkflowsMu: sync.Mutex{},
+		lggr:                  lggr,
+	}, nil
+}
+
+func (c *Capability) Start(ctx context.Context) error {
+	return c.registry.Add(ctx, c)
+}
+
+func (c *Capability) Close() error {
+	return nil
+}
+
+func (c *Capability) Info(ctx context.Context) (capabilities.CapabilityInfo, error) {
+	return capabilityInfo, nil
+}
+
+func getMessageID(req capabilities.CapabilityRequest) (string, error) {
+	if err := validation.ValidateWorkflowOrExecutionID(req.Metadata.WorkflowID); err != nil {
+		return "", fmt.Errorf("workflow ID is invalid: %w", err)
+	}
+	if err := validation.ValidateWorkflowOrExecutionID(req.Metadata.WorkflowExecutionID); err != nil {
+		return "", fmt.Errorf("workflow execution ID is invalid: %w", err)
+	}
+	messageID := []string{
+		req.Metadata.WorkflowID,
+		req.Metadata.WorkflowExecutionID,
+		webcapabilities.MethodWebAPITarget,
+	}
+	return strings.Join(messageID, "/"), nil
+}
+
+func (c *Capability) Execute(ctx context.Context, req capabilities.CapabilityRequest) (capabilities.CapabilityResponse, error) {
+	c.lggr.Debugw("executing http target", "capabilityRequest", req)
+
+	var input Input
+	err := req.Inputs.UnwrapTo(&input)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, err
+	}
+
+	var workflowCfg WorkflowConfig
+	err = req.Config.UnwrapTo(&workflowCfg)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, err
+	}
+
+	messageID, err := getMessageID(req)
+	if err != nil {
+		return capabilities.CapabilityResponse{}, err
+	}
+
+	c.registeredWorkflowsMu.Lock()
+	defer c.registeredWorkflowsMu.Unlock()
```

Comment:
is it necessary to hold the lock for the duration of this method or can it be released after reading the map?

Reply:
yep can release it after reading the map 👍